### DISCUSSION TITLE
DataTable should have a caption prop / slot

### DISCUSSION THREAD

**Discussion by laruiss** [primefaces/community#1583] [PrimeVue]
For a11y, I think it is important that DataTable allow for a caption. Something like this:

```html
                    <table ref="table" role="table" :class="[cx('table'), tableClass]" :style="[tableStyle, slotProps.spacerStyle]" v-bind="{ ...tableProps, ...ptm('table') }">
                        <slot name="caption">
                            <caption v-if="caption" :class="captionClass">
                                {{ caption }}
                            </caption>
                        </slot>
                        <!-- ... -->
```

And if I understand correctly, BaseDataTable should have

```ts
// (...)
   props: {
        // (...)
        caption: {
            type: String,
            default: null
        },
        captionClass: {
            type: [Object, Array, String],
            default: null
        },
        // (...)
  }
// (...)
```

Also, I think it is not a good practice to add `role="table"` to the tag `<table>` which has natively this role, and furthermore, it might be a good idea, for a datatable, to set the role to [`grid`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/grid_role). Even better: this should be customizable through prop.

Thank you for all the amazing work!

**Comment by ewillebrands** [comment]
The caption element should be rendered as a direct child of the table, as a sibling of the columns. In addition to a caption it would also be useful to add a summary as a nested element of the caption, to be able to describe the behaviour of more complex implementations of a dataTable. For example when it has expandable rows, you would be able to prepare visually impaired users before they encounter unexpected behaviour of adding an extra row.
For guidelines on this, have a look at https://www.w3.org/WAI/tutorials/tables/caption-summary/
